python/ray/dag/compiled_dag_node.py

```diff
@@ -858,6 +865,159 @@ def _get_node_id(self):
+            def in_degree(self) -> int:
+                return len(self.in_edges)
+
+        from ray.dag import ClassMethodNode
+
+        def _get_next_task_idx(task: "CompiledTask") -> Optional[int]:
+            if not isinstance(task.dag_node, ClassMethodNode):
+                return None
+            actor_handle = task.dag_node._get_actor_handle()
+            bind_index = task.dag_node._get_bind_index()
+            for same_node_task in self.actor_to_tasks[actor_handle]:
+                if same_node_task.dag_node._get_bind_index() == bind_index + 1:
+                    return same_node_task.idx
+            return None
+
+        def _add_edge(
+            graph: Dict[int, GraphNode], from_idx: int, to_idx: Optional[int]
+        ):
+            if to_idx is None:
+                return
+            graph[from_idx].out_edges.add(to_idx)
+            graph[to_idx].in_edges.add(from_idx)
+
+        graph = defaultdict(GraphNode)
+        for idx, task in self.idx_to_task.items():
+            # Add an edge from task_{bind_index} to task_{bind_index+1}
+            # on the same actor.
+            next_task_idx = _get_next_task_idx(task)
+            _add_edge(graph, idx, next_task_idx)
+            for downstream_idx in task.downstream_node_idxs:
+                if task.dag_node.type_hint.requires_nccl():
+                    # Add an edge from the writer and reader of an NCCL channel
+                    # to the node that has the next bind index on the same actor
+                    # as the writer.
+                    _add_edge(graph, downstream_idx, next_task_idx)
+                else:
+                    # Add an edge from the writer to the reader if the channel
+                    # isn't an NCCL channel.
+                    _add_edge(graph, idx, downstream_idx)
+        num_total_nodes = len(graph)
+
+        # A list of nodes with in-degree 0, including (1) InputNode and
+        # (2) the nodes that only read from NCCL channels and are the first
+        # node on the actor.
+        zero_in_degree_nodes = deque()
+        for idx, node in graph.items():
+            if node.in_degree == 0:
+                zero_in_degree_nodes.append(idx)
+        visited_nodes = set()
+
+        # Perform topological sort to find a topological order of the graph.
+        # If topological order exists, the graph is a DAG. Otherwise, it has
+        # a cycle.
+        while zero_in_degree_nodes:
+            node = zero_in_degree_nodes.popleft()
+            visited_nodes.add(node)
+            for out_node in graph[node].out_edges:
+                graph[out_node].in_edges.remove(node)
+                if graph[out_node].in_degree == 0:
+                    zero_in_degree_nodes.append(out_node)
+
+        # Remove visited nodes from the graph.
+        for node in visited_nodes:
+            del graph[node]
+
+        topological_order_exists = len(visited_nodes) == num_total_nodes
+        if not topological_order_exists:
```

Comment:
@stephanie-wang do you have any suggestions about the error message here?

Reply:
Let's just keep it simple and not try to log a better error message for now. How about just something like this as the Exception message?
```
The compiled DAG may hang due to blocking NCCL calls. If you believe this is a false positive, please file an issue at https://github.com/ray-project/ray/issues/new/
```

Reply:
Updated https://github.com/ray-project/ray/pull/45960/commits/4babe2294f4a95cdb8e2f888727464cab58e72e9